### PR TITLE
Properties substitutions support for resource's parents

### LINKED ISSUE #3222: Support dynamically fetching pipelines values from the resource parents
## Description

As a TRE Developer 
I want to be able to request (inside a porter pipeline) to substitute a value which originates from the resource parents.
So that values such as the parent selected suffix can be dynamically retrieved.

## Acceptance criteria

- [ ] support the syntax of {{ parent_workspace.properties.unique_identifier_suffix }} AND "{{ parent_workspace_service.properties.unique_identifier_suffix }}"
- [ ] Passing UTs

### PR DESCRIPTION
**Pull request by eladiw**
# Resolves #3222

## What is being addressed

Enhancing the already existing properties substitution mechanism so it will be able to fetch properties from the parents as well (for a user resource, it means from the ws and the ws-svc, for a workspace service it means from the workspace).

## How is this addressed

- Updated the substitution mechanism to receive 2 more dictionaries of the 2 parents.
- supporting resource.parent and resource.parent.parent, so we can select the right dictionary to use
- enhanced unit tests
- see screenshot for usage example

image

### CODE REVIEW

File: api_app/service_bus/substitutions.py [hunk 1]
```diff
@@ -89,18 +102,53 @@ def substitute_value(val: str, primary_resource_dict: dict) -> Union[dict, list,
...
+                if p[i] == "parent":
+                    hierarchy_level += 1
+
+        # sanity
+        if primary_resource_type == "user-resource" and hierarchy_level > 2:
+            raise ValueError("parent.parent.parent is invalid for a resource of type 'user-resource'")
+        elif primary_resource_type == "workspace-service" and hierarchy_level > 1:
+            raise ValueError("parent.parent is invalid for a resource of type 'workspace-service'")
+        elif primary_resource_type == "workspace" and hierarchy_level > 0:
+            raise ValueError("parent is invalid for a resource of type 'workspace'")
```
Comment (tamirkamara) [inline api_app/service_bus/substitutions.py:125]: Don't we have enums for these strings?

File: api_app/service_bus/substitutions.py [hunk 2]
```diff
@@ -89,18 +102,53 @@ def substitute_value(val: str, primary_resource_dict: dict) -> Union[dict, list,
...
+            if len(p) > i:
+                if p[i] == "parent":
+                    hierarchy_level += 1
+
+        # sanity
+        if primary_resource_type == "user-resource" and hierarchy_level > 2:
+            raise ValueError("parent.parent.parent is invalid for a resource of type 'user-resource'")
+        elif primary_resource_type == "workspace-service" and hierarchy_level > 1:
+            raise ValueError("parent.parent is invalid for a resource of type 'workspace-service'")
+        elif primary_resource_type == "workspace" and hierarchy_level > 0:
```
Comment (tamirkamara) [inline api_app/service_bus/substitutions.py:124]: What will happen for shared-service?

File: api_app/service_bus/substitutions.py [hunk 3]
```diff
@@ -89,18 +102,53 @@ def substitute_value(val: str, primary_resource_dict: dict) -> Union[dict, list,
             t = p[0:p.index("}}")]
             tokens.append(t)
 
+    dict_to_use = None
     for t in tokens:
-        # t = "resource.properties.prop_1"
+        # t = "{resource/parent_workspace/parent_workspace_service}.properties.prop_1"
```
Comment (tamirkamara) [inline api_app/service_bus/substitutions.py:107]: Is the order correct?

**Comment by eladiw** [comment]
@tamirkamara  fixed.

File: api_app/service_bus/substitutions.py [hunk 4]
```diff
@@ -89,18 +103,54 @@ def substitute_value(val: str, primary_resource_dict: dict) -> Union[dict, list,
             t = p[0:p.index("}}")]
             tokens.append(t)
 
+    dict_to_use = None
     for t in tokens:
-        # t = "resource.properties.prop_1"
+        # t = "{resource[.parent][.parent].properties.prop_1"
```
Comment (anatbal) [inline api_app/service_bus/substitutions.py:108]: why is this commented out?

File: api_app/service_bus/substitutions.py [hunk 5]
```diff
@@ -89,18 +103,54 @@ def substitute_value(val: str, primary_resource_dict: dict) -> Union[dict, list,
...
+            elif primary_resource_type == strings.RESOURCE_TYPE_WORKSPACE_SERVICE:
+                dict_to_use = primary_parent_ws_dict
+            del p[1]
+        else:
+            dict_to_use = primary_resource_dict
+
+        prop_to_get = dict_to_use
+        for i in range(1, len(p)):
+            # instead of failing, if the value is not found, return empty string. Used for backward compatability
+            if p[i] not in prop_to_get:
```
Comment (anatbal) [inline api_app/service_bus/substitutions.py:146]: nit: I would suggest for readability put p[i] in a var since it is being used multiple time

File: api_app/service_bus/substitutions.py [hunk 6]
```diff
@@ -89,18 +103,54 @@ def substitute_value(val: str, primary_resource_dict: dict) -> Union[dict, list,
...
+            raise ValueError("parent.parent is invalid for a resource of type '{}'".format(str(strings.RESOURCE_TYPE_WORKSPACE_SERVICE)))
+        elif primary_resource_type == strings.RESOURCE_TYPE_WORKSPACE and hierarchy_level > 0:
+            raise ValueError("parent is invalid for a resource of type '{}'".format(str(strings.RESOURCE_TYPE_WORKSPACE)))
+        elif primary_resource_type == strings.RESOURCE_TYPE_SHARED_SERVICE and hierarchy_level > 0:
+            raise ValueError("parent is invalid for a resource of type '{}'".format(str(strings.RESOURCE_TYPE_SHARED_SERVICE)))
+
+        if hierarchy_level == 2:
+            if primary_resource_type == strings.USER_RESOURCE:
+                dict_to_use = primary_parent_ws_dict
+                del p[2]
```
Comment (anatbal) [inline api_app/service_bus/substitutions.py:132]: can we change p, t to names with more meaning since this code is already a little bit hard to follow

**Review by anatbal — APPROVED** [review]
LGTM, added minor comments
Of course, assuming the cascade is still working

**Comment by anatbal** [comment]
BTW - test is not enough in this case. Do an extended test

File: api_app/service_bus/substitutions.py (hunk 4 above)
Comment (eladiw) [reply api_app/service_bus/substitutions.py:108]: The comment says this is an example of how prop substitute looks like.
I added that 2 '.parent' are also an option

File: api_app/_version.py [hunk 7]
```diff
@@ -1 +1 @@
-__version__ = "0.12.0"
+__version__ = "0.12.1"
```
Comment (tamirkamara) [inline api_app/_version.py:1]:
```suggestion
__version__ = "0.13.0"
```

**Review by tamirkamara — APPROVED** [review]

File: api_app/service_bus/helpers.py [hunk 8]
```diff
@@ -109,22 +122,23 @@ async def try_upgrade_with_retries(num_retries: int, attempt_count: int, resourc
                 user=user,
                 resource_to_update_id=resource_to_update_id,
                 template_step=template_step,
-                primary_resource=primary_resource
+                primary_resource=primary_resource,
+                primary_parent_workspace=primary_parent_workspace
```
Comment (anatbal) [inline api_app/service_bus/helpers.py:126]: add  primary_parent_workspace_svc=primary_parent_workspace_svc

**Review by anatbal — APPROVED** [review]
